src/genworld.cpp

```diff
@@ -337,3 +341,157 @@ void GenerateWorld(GenWorldMode mode, uint size_x, uint size_y, bool reset_setti
+ */
+void LoadTownData()
+{
+	/* Load the JSON file as a string initially. We'll parse it soon. */
+	size_t filesize;
+	FILE *f = FioFOpenFile(_file_to_saveload.name, "rb", HEIGHTMAP_DIR, &filesize);
+
+	if (f == nullptr) {
+		ShowErrorMessage(STR_TOWN_DATA_ERROR_LOAD_FAILED, STR_TOWN_DATA_ERROR_JSON_FORMATTED_INCORRECTLY, WL_ERROR);
+		return;
+	}
+
+	std::string text(filesize, '\0');
+	size_t len = fread(text.data(), filesize, 1, f);
+	FioFCloseFile(f);
+	if (len != 1) {
+		ShowErrorMessage(STR_TOWN_DATA_ERROR_LOAD_FAILED, STR_TOWN_DATA_ERROR_JSON_FORMATTED_INCORRECTLY, WL_ERROR);
+		return;
+	}
+
+	/* Now parse the JSON. */
+	nlohmann::json town_data;
+	try {
+		town_data = nlohmann::json::parse(text);
+	} catch (nlohmann::json::exception &) {
+		ShowErrorMessage(STR_TOWN_DATA_ERROR_LOAD_FAILED, STR_TOWN_DATA_ERROR_JSON_FORMATTED_INCORRECTLY, WL_ERROR);
+		return;
+	}
+
+	/* Check for JSON formatting errors with the array of towns. */
+	if (!town_data.is_array()) {
+		ShowErrorMessage(STR_TOWN_DATA_ERROR_LOAD_FAILED, STR_TOWN_DATA_ERROR_JSON_FORMATTED_INCORRECTLY, WL_ERROR);
+		return;
+	}
+
+	std::vector<std::pair<Town *, uint> > towns;
+	uint failed_towns = 0;
+
+	/* Iterate through towns and attempt to found them. */
+	for (auto &feature : town_data) {
+		ExternalTownData town;
+
+		/* Ensure JSON is formatted properly. */
+		if (!feature.is_object()) {
+			ShowErrorMessage(STR_TOWN_DATA_ERROR_LOAD_FAILED, STR_TOWN_DATA_ERROR_JSON_FORMATTED_INCORRECTLY, WL_ERROR);
+			return;
+		}
+
+		/* Check to ensure all fields exist and are of the correct type.
+		 * If the town name is formatted wrong, all we can do is give a general warning. */
+		if (!feature.contains("name") || !feature.at("name").is_string()) {
+			ShowErrorMessage(STR_TOWN_DATA_ERROR_LOAD_FAILED, STR_TOWN_DATA_ERROR_JSON_FORMATTED_INCORRECTLY, WL_ERROR);
+			return;
+		}
+
+		/* If other fields are formatted wrong, we can actually inform the player which town is the problem. */
+		if (!feature.contains("population") || !feature.at("population").is_number() ||
+				!feature.contains("city") || !feature.at("city").is_boolean() ||
+				!feature.contains("x") || !feature.at("x").is_number() ||
+				!feature.contains("y") || !feature.at("y").is_number()) {
+			feature.at("name").get_to(town.name);
+			SetDParamStr(0, town.name);
+			ShowErrorMessage(STR_TOWN_DATA_ERROR_LOAD_FAILED, STR_TOWN_DATA_ERROR_TOWN_FORMATTED_INCORRECTLY, WL_ERROR);
+			return;
+		}
+
+		/* Set town properties. */
+		feature.at("name").get_to(town.name);
+		feature.at("population").get_to(town.population);
+		feature.at("city").get_to(town.is_city);
+
+		/* Set town coordinates. */
+		feature.at("x").get_to(town.x_proportion);
+		feature.at("y").get_to(town.y_proportion);
+
+		/* Check for improper coordinates and warn the player. */
+		if (town.x_proportion <= 0 || town.y_proportion <= 0 || town.x_proportion >= 1 || town.y_proportion >= 1) {
```

Comment:
```suggestion
		if (town.x_proportion <= 0.0 || town.y_proportion <= 0.0 || town.x_proportion >= 1.0 || town.y_proportion >= 1.0) {
```

Don't confuse your reviewers :P